### DISCUSSION TITLE
Various improvement suggestions - invalid html - long attributes

### DISCUSSION THREAD

**Discussion by Robin-Wils** [felix-berlin/astro-breadcrumbs#418] [Ideas]
I have noticed that this component generates invalid html. The tag `astro-breadcrumbs` is not valid html. Everything still worked when I moved the `nav` element out of that tag. Aside from that it uses a few long attributes.

**Invalid HTML**
Can the invalid tag `astro-breadcrumbs` be replaced or removed? Replacing it with a `div` fixes this already.

**Compactness**
This component isn't as tiny as I would like. The `astro-breadcrumbs` tag includes many data-elements like `id`. The `id` attribute is filled even if I do not fill it in. The value that gets inserted is long. While this is a minor thing, improving this would be beneficial for users who value smaller HTML payloads and faster load times.

**Describe alternatives you've considered**
I am considering reverting back to my own custom breadcrumbs component which I used before. While my custom implementation already supports schema.org and related features, it lacks advanced features like truncating breadcrumbs. It is mostly the invalid html that is making me consider this.

I would still prefer a package like this. If the pros outweigh the cons. It has less maintenance cost for me (the user). Thanks for creating this component.

**Comment by felix-berlin** [comment]
Hi @Robin-Wils,
thanks for your detailed issue.

You are right, the `<astro-breadcrumbs>` component is not a native HTML element, which does not mean that it has to be invalid HTML.

**Regarding invalid HTML**
The component is a web component which allows you to create your own custom element.

**Regarding Compactness**
All `data-attributes` are required to pass certain data to the clientside JS where it is needed. It is therefore difficult to become more minimalistic here. Only the data that is really needed is transmitted.

All in all, I don't know how I can help you at this point. If you have specific suggestions for improvement, please send me a pull request.

By the way, Astro Islands are also web components. Here is the [corresponding Docs page](https://docs.astro.build/en/guides/client-side-scripts/#pass-frontmatter-variables-to-scripts) on the subject.

![image](https://github.com/user-attachments/assets/61e1da99-6590-428f-a950-e86a3c8942be)

[MDN Web-Component Docs](https://developer.mozilla.org/en-US/docs/Web/API/Web_components)

**Comment by Robin-Wils** [comment]
Hmm, thanks for the detailed reply. I have worked with similar web components before (I think in vue, react and angular). But it only occurred to me now that they use custom elements. 

Although, I now concluded for myself that I do not like the idea of web components. It seems overengineered, makes the html bigger and the names can confuse people. I likely will write my own tools. The YouTube embed I use - uses web components too. Maintaining your own tools does have big disadvantages too though, mainly maintenance.
